sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala

```diff
@@ -463,6 +463,8 @@ object SimplifyConditionals extends Rule[LogicalPlan] with PredicateHelper {
       case If(Literal(null, _), _, falseValue) => falseValue
       case If(cond, trueValue, falseValue)
         if cond.deterministic && trueValue.semanticEquals(falseValue) => trueValue
+      case If(p, l @ Literal(null, _), FalseLiteral) if !p.nullable => And(p, l)
+      case If(p, l @ Literal(null, _), TrueLiteral) if !p.nullable => Or(Not(p), l)
```

Comment:
Hm, I think for expression rewriting, this won't be much faster. This is like @HyukjinKwon said, should be a rare case. The most likely usage, I think, is to turn a predicate to a form being able to push down to data source.

Adding a rule is too much. As part of existing rule looks more proper.

Reply:
Hmm, I'm rethinking this. Is a predicate `And(p, null)` be pushed down? Predicate pushdown won't rewrite null literals.

Reply:
I'm yet to find a case where `If(p, null, false)` gets populated to datasource. I think most of the time they just get replaced by `false` so transforming them to `And(p, null)` doesn't seem to help much.

On the other hand, `Or(Not(p), null)` does get pushed down to datasources.

Reply:
Pushdown predicate is somehow different to general predicate. For example, IIUC we won't push down `Or(Not(p), null)` because of the null literal. Predicate pushdown only rewrites predicates applied to a field column, e.g. col > 1.

Just for predicate pushdown, maybe we can transfer `Or(Not(p), null)` to just `Not(p)`? Because if `p` is true, the predicate evaluates to null, then we filter out the row. If `p` is false, the predicate value is true. So it semantically equals to `Not(p)` for predicate used in pushdown.

I'm not sure if I miss anything or edge case, and this also doesn't work for nested predicate. Besides, this is unlikely case too I think, we can consider if it's worth the changing.

Reply:
Hmm, I'm not sure why `Or(Not(p), null)` can't be pushed down. I have this test case:
```scala
  test("test pushdown if") {
    withTempPath( dir => {
      spark.range(1, 100).toDF("c").write.parquet(dir.getAbsolutePath)
      withTempView("t") {
        spark.read.parquet(dir.getAbsolutePath).createOrReplaceTempView("t")
        sql("SELECT * FROM t WHERE if(isnull(c), null, true)").explain()
      }
    })
  }
```

Without the rule, the explain result is:
```
*(1) Filter if (isnull(id#223L)) false else true
+- *(1) ColumnarToRow
   +- FileScan parquet [id#223L] Batched: true, DataFilters: [if (isnull(id#223L)) false else true], Format: Parquet, Location: InMemoryFileIndex[file:/private/var/folders/z3/ptkgr4kn4pv9v8g7s1fnr5mm0000gn/T/spark-858a579d-3a..., PartitionFilters: [], PushedFilters: [], ReadSchema: struct<id:bigint>
```

with the rule it become:
```
== Physical Plan ==
*(1) Filter isnotnull(id#223L)
+- *(1) ColumnarToRow
   +- FileScan parquet [id#223L] Batched: true, DataFilters: [isnotnull(id#223L)], Format: Parquet, Location: InMemoryFileIndex[file:/private/var/folders/z3/ptkgr4kn4pv9v8g7s1fnr5mm0000gn/T/spark-08e9254e-df..., PartitionFilters: [], PushedFilters: [IsNotNull(id)], ReadSchema: struct<id:bigint>
```

so you can see the `Not(p)` here is pushed down to data sources. Did I miss anything?

Reply:
Oh, I see.

```
== Analyzed Logical Plan ==
c: bigint
Project [c#5849L]
+- Filter if (isnull(c#5849L)) cast(null as boolean) else true
   +- SubqueryAlias t
      +- Relation[c#5849L] parquet

== Optimized Logical Plan ==
Filter isnotnull(c#5849L)
+- Relation[c#5849L] parquet
```

The predicate actually will be optimized to a simple `IsNotNull`.

Reply:
> Just for predicate pushdown, maybe we can transfer Or(Not(p), null) to just Not(p)? 

Then `Or(Not(p), null)` is already optimized to `Not(p)` now for predicate pushdown. But this is only for non-nested predicate case.

Reply:
Thanks for being patient with me. I think this new pattern looks okay.